test/pickfirst_leaf_test.go

```diff
@@ -0,0 +1,607 @@
+				activeAddrs = append(activeAddrs, addrs[idx])
+			}
+			r.UpdateState(resolver.State{Addresses: activeAddrs})
+			bal := <-balChan
+			select {
+			case <-bal.resolverUpdateSeen:
+			case <-ctx.Done():
+				t.Fatalf("Context timed out waiting for resolve update to be processed")
+			}
+
+			// shutdown all active backends except the target.
+			var targetAddr resolver.Address
+			for idxI, idx := range tc.preUpdateBackendIndexes {
+				if idx == tc.preUpdateTargetBackendIndex {
+					targetAddr = activeAddrs[idxI]
+					continue
+				}
+				backends[idx].S.Stop()
+			}
+
+			if err := pickfirst.CheckRPCsToBackend(ctx, cc, targetAddr); err != nil {
+				t.Fatal(err)
+			}
+
+			if diff := cmp.Diff(scStateExpectationToScState(tc.wantPreUpdateScStates, addrs), bal.subConns()); diff != "" {
+				t.Errorf("subconn states mismatch (-want +got):\n%s", diff)
+			}
+
+			if len(tc.updatedBackendIndexes) == 0 {
+				if diff := cmp.Diff(tc.wantConnStateTransitions, bal.connStateTransitions()); diff != "" {
+					t.Errorf("balancer states mismatch (-want +got):\n%s", diff)
+				}
+				return
+			}
+
+			// Restart all the backends.
+			for i, s := range backends {
+				if !tc.restartConnected && i == tc.preUpdateTargetBackendIndex {
+					continue
+				}
+				s.S.Stop()
+				if err := s.StartServer(); err != nil {
+					t.Fatalf("Failed to re-start test backend: %v", err)
+				}
+			}
+
+			activeAddrs = []resolver.Address{}
+			for _, idx := range tc.updatedBackendIndexes {
+				activeAddrs = append(activeAddrs, addrs[idx])
+			}
+
+			r.UpdateState(resolver.State{Addresses: activeAddrs})
+			select {
+			case <-bal.resolverUpdateSeen:
+			case <-ctx.Done():
+				t.Fatalf("Context timed out waiting for resolve update to be processed")
+			}
+
+			// shutdown all active backends except the target.
+			for idxI, idx := range tc.updatedBackendIndexes {
+				if idx == tc.postUpdateTargetBackendIndex {
+					targetAddr = activeAddrs[idxI]
+					continue
+				}
+				backends[idx].S.Stop()
+			}
+
+			if err := pickfirst.CheckRPCsToBackend(ctx, cc, targetAddr); err != nil {
+				t.Fatal(err)
+			}
+
+			if diff := cmp.Diff(scStateExpectationToScState(tc.wantPostUpdateScStates, addrs), bal.subConns()); diff != "" {
+				t.Errorf("subconn states mismatch (-want +got):\n%s", diff)
+			}
+
+			if diff := cmp.Diff(tc.wantConnStateTransitions, bal.connStateTransitions()); diff != "" {
+				t.Errorf("balancer states mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func (s) TestPickFirstLeaf_EmptyAddressList(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	balChan := make(chan *stateStoringBalancer, 1)
+	balancer.Register(&stateStoringBalancerBuilder{balancerChan: balChan})
+	cc, r, backends := setupPickFirstLeaf(t, 1)
+	addrs := stubBackendsToResolverAddrs(backends)
+
+	r.UpdateState(resolver.State{Addresses: addrs})
+	bal := <-balChan
+	select {
+	case <-bal.resolverUpdateSeen:
```

Comment:
We have a `testutils.AwaitState` helper function which we could use instead of this state storing balancer. The approach would be as follows:
- Configure `pickfirstleaf` as the LB policy on the channel instead of the wrapper LB policy
- Push an update with addresses
- Use `testutils.AwaitState` and wait for the channel to become `Ready`
- Mark an RPC and verify that it reaches the first backend
- Psh an update with no addresses
- Use `testutils.AwaitState` and wait for the channel to become `TF`
- Push an update with addresses
- Use `testutils.AwaitState` and wait for the channel to become `Ready`
- Mark an RPC and verify that it reaches the first backend

Let me know what you think about this approach.

And for verifying the sequence of state transitions, we have an internal API: https://github.com/grpc/grpc-go/blob/e4b09f111dd3856bc1a3ed4b234edbb5d3a861c1/internal/internal.go#L139
This is also rather new. So, you wont find a lot of usages in our codebase. But I'm trying to encourage new tests to use this. Please let me know if you have any questions on this. Thanks.